Mathlib/Control/Random.lean

```diff
@@ -0,0 +1,145 @@
+
+namespace Rng
+
+def range : Rng → Nat × Nat
+| stdGen gen => RandomGen.range gen
+
+def next : Rng → Nat × Rng
+| stdGen gen => RandomGen.next gen |>.map id stdGen
+
+def split : Rng → Rng × Rng
+| stdGen gen => RandomGen.split gen |>.map stdGen stdGen
+
+instance : RandomGen Rng := ⟨range, next, split⟩
+
+end Rng
+
+namespace Rand
+  /-- Generate one more `Nat` -/
+  def next {g : Type} [RandomGen g] : RandG g Nat := do
+    let rng := (←get).down
+    let (res, new) := RandomGen.next rng
+    set (ULift.up new)
+    pure res
+
+  /-- Create a new random number generator distinct from the one stored in the state -/
+  def split {g : Type} [RandomGen g] : RandG g g := do
+    let rng := (←get).down
+    let (r1, r2) := RandomGen.split rng
+    set (ULift.up r1)
+    pure r2
+
+  /-- Get the range of Nat that can be generated by the generator `g` -/
+  def range {g : Type} [RandomGen g] : RandG g (Nat × Nat) := do
+    let rng := (←get).down
+    pure <| RandomGen.range rng
+end Rand
+
+namespace Random
+
+open Rand
+
+/-- Generate a random value of type `α`. -/
+def rand (α : Type u) [Random α] [RandomGen g] : RandG g α := Random.random
+
+/-- Generate a random value of type `α` between `x` and `y` inclusive. -/
+def randBound (α : Type u) [Preorder α]  [BoundedRandom α] (lo hi : α) (h : lo ≤ hi) [RandomGen g] : RandG g {a // lo ≤ a ∧ a ≤ hi} :=
+  BoundedRandom.randomR lo hi h
+
+def randFin {n : Nat} [RandomGen g] : RandG g (Fin n.succ) :=
+  λ ⟨g⟩ => randNat g 0 n.succ |>.map Fin.ofNat ULift.up
+
+instance {n : Nat} : Random (Fin n.succ) where
+  random := randFin
+
+def randBool [RandomGen g] : RandG g Bool := do
+  let fin ← rand (Fin 1)
+  if fin == 1 then
+    pure true
+  else
+    pure false
+
+instance : Random Bool where
+  random := randBool
+
+instance : BoundedRandom Nat where
+  randomR := λ lo hi h _ => do
+    let z ← rand (Fin (hi - lo).succ)
+    pure ⟨
+      lo + z.val, Nat.le_add_right _ _,
+      by
+         apply Nat.add_le_of_le_sub_left
+         exact h
+         apply Nat.le_of_succ_le_succ
+         exact z.isLt
+    ⟩
+
+instance : BoundedRandom Int where
+  randomR := λ lo hi h _ => do
+    let ⟨z, h1, h2⟩ ← randBound Nat 0 (Int.natAbs $ hi - lo) (Nat.zero_le _)
+    pure ⟨
+      z + lo,
+      Int.le_add_of_nonneg_left (Int.ofNat_zero_le z),
+      Int.add_le_of_le_sub_right $ Int.le_trans
+        (Int.ofNat_le.mpr h2)
+        (by
+          apply le_of_eq
+          apply Int.ofNat_natAbs_eq_of_nonneg
+          exact Int.sub_nonneg_of_le h)
+    ⟩
+
+instance {n : Nat} : BoundedRandom (Fin n) where
+  randomR := λ lo hi h _ => do
+    let ⟨r, h1, h2⟩ ← randBound Nat lo.val hi.val h
+    pure ⟨⟨r, Nat.lt_of_le_of_lt h2 hi.isLt⟩, h1, h2⟩
+
+end Random
+
+
+def IO.runRand (cmd : Rand α) : IO α := do
+  let stdGen ← stdGenRef.get
+  let rng := ULift.up <| Rng.stdGen stdGen
+  pure <| Prod.fst <| Id.run <| StateT.run cmd rng
```

Comment:
This should set the `stdGenRef` afterwards to the updated value.

I wonder if it would be more useful to have an `abbrev RandT (g : Type) := StateT (ULift g)` and then a `def IO.runRand [MonadLiftT BaseIO m] (cmd : RandT StdGen m α) : m α`.

There's also two big issues with the `stdGenRef` (which should at least be document):
1. It is not thread-local (meaning two threads that call `runRand` at the same time will get the same generator).
2. It is not necessarily initialized (i.e., `def main := do IO.runRand ..` is deterministic).